cps/kni.c

```diff
@@ -1105,6 +1230,96 @@ rd_getroute(const struct nlmsghdr *req, struct cps_config *cps_conf, int *err)
 	return MNL_CB_OK;
 }
 
+static void
+rd_fill_getlink_reply(struct cps_config *cps_conf,
+	struct mnl_nlmsg_batch *batch,
+	const char *kni_name, unsigned int kni_index, unsigned int kni_mtu,
+	uint32_t seq)
+{
+	struct nlmsghdr *reply;
+	struct ifinfomsg *ifim;
+
+	reply = mnl_nlmsg_put_header(mnl_nlmsg_batch_current(batch));
+	reply->nlmsg_type = RTM_NEWLINK;
+	reply->nlmsg_flags = NLM_F_MULTI;
+	reply->nlmsg_seq = seq;
+	reply->nlmsg_pid = cps_conf->nl_pid;
+
+	ifim = mnl_nlmsg_put_extra_header(reply, sizeof(*ifim));
+	ifim->ifi_family = AF_UNSPEC;
+	ifim->ifi_type = ARPHRD_ETHER;
+	ifim->ifi_index = kni_index;
+	ifim->ifi_flags = 0;
+	ifim->ifi_change = 0xFFFFFFFF;
+
+	mnl_attr_put_str(reply, IFLA_IFNAME, kni_name);
+	mnl_attr_put_u32(reply, IFLA_MTU, kni_mtu);
+}
+
+static inline uint16_t
+get_min_mtu(struct gatekeeper_if *iface, uint16_t kni_mtu)
+{
+	return iface->mtu < kni_mtu ? iface->mtu : kni_mtu;
+}
+
+static int
+rd_getlink(const struct nlmsghdr *req, struct cps_config *cps_conf, int *err)
+{
+	char buf[2 * MNL_SOCKET_BUFFER_SIZE];
+	struct mnl_nlmsg_batch *batch;
+
+	batch = mnl_nlmsg_batch_start(buf, MNL_SOCKET_BUFFER_SIZE);
+	if (batch == NULL) {
+		CPS_LOG(ERR, "Failed to allocate a batch for a GETLINK reply\n");
+		*err = -ENOMEM;
+		goto out;
+	}
+
+	rd_fill_getlink_reply(cps_conf, batch,
+		rte_kni_get_name(cps_conf->front_kni),
+		cps_conf->front_kni_index,
+		get_min_mtu(&cps_conf->net->front, cps_conf->front_kni_mtu),
+		req->nlmsg_seq);
+	if (!mnl_nlmsg_batch_next(batch)) {
+		/*
+		 * Couldn't add link information for front;
+		 * report error to routing daemon.
+		 */
+		CPS_LOG(ERR, "Failed to add link information about front interface in response to a GETLINK request\n");
+		*err = -ENOMEM;
+		goto free_batch;
```

Comment:
Same explanation as the comment on the other patch. If this call fails, it means the message wasn't actually added to the batch (because room ran out) and therefore there's nothing to send.

Realistically, these conditions should never happen. We just created the batch, and all of the messages we're sending should be much less than `MNL_SOCKET_BUFFER_SIZE` because there's only two entries at most.

Reply:
You have skipped this item.

The code would be like this:
```
*err = rd_send_batch(cps_conf, batch, "LINK", req->nlmsg_seq, req->nlmsg_pid, false);
if (*err < 0)
	goto free_batch;
```

Again, this comment impacts code below for the back interface AND the next patch since it uses the same code pattern.